File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    public int maxSubArray(int[] arr) {
        int n = arr.length;
        
        int sum = 0;
        int ans = Integer.MIN_VALUE;

        for(int i=0;i<n;i++)
        {
            if(sum>=0)
                sum+=arr[i];
            else
                sum = arr[i];

            ans = Math.max(sum,ans);
        }

        return ans;

    }
}

File: 53-maximum-subarray/maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {

        //kadanes algorithm
        // ansStart and ansEnd are the indexes of the subarray formed

        long long maxi = INT_MIN;
        long long sum = 0;
        for(int i=0;i<nums.size();i++)
        {

            sum+=nums[i];

            if(sum>maxi)
            {
                maxi = sum;
            }

            if(sum<0)
            {
                sum = 0;
            }
        }

        return maxi;
    }
};